### ISSUE TITLE
Osara move to chord actions generating MIDI editor: Unselect all events undo points

### ISSUE THREAD

**Issue by ns-studios** [jcsteh/osara#1088]
I don't know if this has been talked about before, or if there's even anything Osara can do about it, but using actions such as
OSARA: Move to next chord
OSARA: Move to previous chord
OSARA: Move to previous note in chord
OSARA: Move to next note in chord
in the midi editor seems to generate
MIDI editor: Unselect all events
undo point every time they're used which then get in the way when trying to undo an actual action like move notes up one semitone, because you have to undo the unselect all events actions first.
Ways to reproduce:
open midi editor, and select piano role
Item: Open in built-in MIDI editor (set default behavior in preferences)
Mode: Piano Roll
left/right/up/down arrow around the notes
OSARA: Move to previous chord
OSARA: Move to next chord
OSARA: Move to previous note in chord
OSARA: Move to next note in chord
then undo, and you should get at least one
MIDI editor: Unselect all events
point wen undoing.
Sometimes only one unselect undo point is made after first using left/right arrow, and then another after editing one of the notes, but sometimes every use of arrows seems to generate an undo point, which can get really annoying to undo.
Would it be possible for Osara not to generate undo points for unselect all unless the action is performed manually?

**Comment by ScottChesworth** [comment]
I'm adding some undo blocks now, giving those points friendlier names that should also work well in other languages (think it'll happen automatically as the action names are already translated).
IMO removing undo points around navigation would need some discussion of the tradeoffs, seeing as navigation is tied to selection. Also... er... I don't know how that would be done yet.

**Comment by ns-studios** [comment]
@ScottChesworth
I've just tried the build with the changes.
I don't think this is much better. I now get an undo point for every osara chord action.
The previous behavior with the unselect all events did exactly that at worst, but sometimes I would only get the unselect all point when using the chord osara action for the first time when opening the midi editor, and then the following actions wouldn't generate any more undo points until I edited the note and then another one would be generated.
I'm even confused why the unselecting would be happening as I'm not manually selecting events, just navigating through them.

**Comment by pitermach** [comment]
In Reaper's general preferences you can exclude various kinds of selection from the undo history and one of the first things I do on a new install is turn most of them off (item navigation in particular gets very spammy). I agree with @ns-studios that ideally for my use case I'd just like the chord navigation to not spam the undo history, I found myself having to spam undos on occasion or review the history.

If there are people who like the current behavior, then at least conceptually it could be tied to one of these undo preferences to make both sides happy, just not sure which one since these aren't Reaper commands and there are no options specific to the midi editor.

**Comment by ScottChesworth** [comment]
Hmm I think something can be done here, but not sure what yet. Agreed that the approach I took earlier isn't it. Will keep experimenting.

**Comment by jcsteh** [comment]
> I'm even confused why the unselecting would be happening as I'm not manually selecting events, just navigating through them.

Navigating through chords or notes selects them, so we have to unselect them when navigating again unless you're doing multiple selection.

You can compress multiple undo points into a single, custom undo point, but i don't know of a way from the API to completely prevent one of REAPER's own undo points. The only thing I can't recall whether I've tried is passing an empty string to Undo_EndBlock, but I very much doubt that'll do it.

**Comment by XHG20033** [comment]
I think the "Unselect all events" undo point is unnecessary most of the time, because when we press the cursor keys, their function is to browse the contents of the MIDI editor, not to change it. Since the event has not changed there is no need to create an undo point.
However, OSARA may need to Unselect all events in the process of browsing notes, but I personally think that at the user level, this step does not need to be presented in most cases. As other friends said, undoing previously made changes requires undoing several "Unselect all events", which is obviously not a good experience.
Or similar to REAPER, provide an option in the settings for users to choose whether to create undo points.

**Comment by ScottChesworth** [comment]
> I can't recall whether I've tried passing an empty string to Undo_EndBlock, but I very much doubt that'll do it.
> 
Tried that just now and nope.

**Comment by jcsteh** [comment]
> I think the "Unselect all events" undo point is unnecessary most of the time, because when we press the cursor keys, their function is to browse the contents of the MIDI editor, not to change it.

That's understood. The key point is that we aren't intentionally generating these undo points. REAPER is and we can't prevent that.

We could try unselecting each selected event ourselves using API calls, rather than executing a REAPER action to do this. My concern is that this might be slow if there are a lot of events selected, whereas REAPER can probably do this faster internally, but I guess it's worth a shot.

**Comment by ScottChesworth** [comment]
Shall I poke Justin before we refactor things, just in case there's some trick we don't know about using UndoBlock?

**Comment by jcsteh** [comment]
UP to you. I'd say there's a 99.9% chance we're not missing something, but I guess it can't hurt to ask if you haven't used up your Justin credits this month. :)

**Comment by ScottChesworth** [comment]
Done. Will report back when I hear back.

**Comment by ScottChesworth** [comment]
Yeah we didn't miss anything. Justin says he'll have a think about a way to prevent selection state being included in undo, says he might be able to add an option.

**Comment by ns-studios** [comment]
That would be cool if there could be an option to not add midi editor selection actions to undo just like Peter mentioned there are options for item/track/envelope selection.
Thank you so much for considering this and even forwarding it to the Reaper devs.

> Yeah we didn't miss anything. Justin says he'll have a think about a way to prevent selection state being included in undo, says he might be able to add an option.

**Comment by ScottChesworth** [comment]
So in today's dev build of REAPER, there's a new Undo category in Prefs, and MIDI events is one of the things that can be unchecked. However, right now it doesn't prevent those unselect undo points from showing up. Hmmm. Not sure where to go from here.

**Comment by ScottChesworth** [comment]
I've followed up and asked for "Edit: Unselect all" to not generate an undo point. Suppose Justin wants to keep this behaviour though, I'm thinking a new "Edit: Unselect notes" action would meet our needs so long as it was included in the new option, right?

**Comment by jcsteh** [comment]
I'm slightly concerned that inserting events and lengthening notes might be part of the new undo setting. Are they?

Edit: Unselect notes would be a change in behaviour I think. If you select all, then press right arrow, you kinda expect that pressing delete will only delete the note you just selected. With unselect notes, I think that will end up deleting any CCs that got selected. Though I'm not entirely sure how select all and unselect all behave with respect to CCs, so maybe I'm wrong.

**Comment by ScottChesworth** [comment]
> I'm slightly concerned that inserting events and lengthening notes might be part of the new undo setting. Are they?
> 
No. I was clear that we only wanted it effecting selection. Have just verified that inserting and lengthening are both still fine.

**Comment by RDMurray** [comment]
I just tried the new setting in Reaper 7.30 on Windows and it is no longer creating extra undo points.

**Comment by ns-studios** [comment]
I tried unchecking midi events from general/undo, but inserting some notes and navigating through chords with OSARA: Move to previous/next chord still generates undo events here with Reaper 7.30 and Osara 2025.1.9.1647,59e8313e.
Is there something I'm missing?

**Comment by RDMurray** [comment]
@ns-studios I don't know, with midi events unchecked I don't see any undo points when navigating in the midi editor on Reaper 7.30. Osara version shouldn't make any difference.

Is anyone else still seeing undo points when midi events is unchecked in undo settings?

**Comment by pitermach** [comment]
Not here. Reaper 7.30 on Mac with a relatively recent OSARA the only events that were in my undo history were the manual note insertions to make the chords for testing.

**Comment by ScottChesworth** [comment]
Think I've just found out why some folks are seeing undo points and others aren't. If you insert notes using actions on I or Shift+I then navigate those, there aren't any unexpected undo points here. However when I navigate MIDI data recorded in using a controller, I'm still getting the extra undo points. I don't have a clue why the results are different.
Can anyone confirm that?

**Comment by ScottChesworth** [comment]
So the new REAPER preference applies to selection, but not unselection, and there hasn't been any more movement on that. I'm still seeing undo points here and still receiving feedback about it through other channels, so I've spent some time trying to prevent it tonight. Wrote a new function that unselects all notes without calling REAPER's unselect all action, but with a lot of MIDI data, it's noticeably slower even on a decent CPU and plenty of fast RAM, so I don't think that's worth pursuing.
Could undo masking like @jcsteh does in the moveToTrack function help with this in some way? I couldn't find details of how the bits work in the API docs, haven't been able to try using that yet.

**Comment by jcsteh** [comment]
That is just about checking whether we want to make our own undo point or not. In this case, REAPER is the one making the undo point. We can merge multiple undo points into one, but we can't avoid a REAPER generated undo point altogether. That is something only REAPER can control.

**Comment by jcsteh** [comment]
> Wrote a new function that unselects all notes without calling REAPER's unselect all action, but with a lot of MIDI data, it's noticeably slower

You could try PreventUIRefresh(1) before the unselection code and PreventUIRefresh(-1) after it. You may also need to call MIDI_DisableSort before, pass bTrue for noSort to the MIDI_Set functions and call MIDI_Sort after all of them, though I don't think this should be necessary (and you shouldn't do it if it isn't) because we're not changing anything that should change ordering.

**Comment by ScottChesworth** [comment]
Aha I hadn't thought of PreventUIRefresh, thanks, will try that.
If it doesn't help, I'm gonna nudge Cockos again and ask for a new action, something like "Edit: Unselect all (no undo). They did insert note (no advance cursor) for us before, maybe this can happen to.

**Comment by jcsteh** [comment]
Any idea why they're pushing back on the undo setting not applying to unselect? If it's just a bug, I don't know that adding a new action will get any more traction.

**Comment by ScottChesworth** [comment]
Dunno. Seems sometimes a new action can happen easier than changing the behaviour of existing ones. I suppose there are scripts and other extensions using them, maybe that's it.
Anyway it doesn't look like PreventUIRefresh makes a difference here, so I've tried again, floated two ways forward, hopefully one of them will land this time.

**Comment by ScottChesworth** [comment]
Cockos are gonna sort this after 7.39 is out. I think that means unselect all will respect the MIDI events preference in the Undo category. Keep an eye on dev builds soon after if you use those, or sit tight if not, hopefully it'll be in 7.40.

**Comment by ScottChesworth** [comment]
Feel free to give today's REAPER dev a try (May 15th). With Prefs -> Undo -> MIDI events unchecked, I'm not seeing extraneous undo events from navigating chords here.
Note that we can't provide any support for REAPER pre-release builds here. If you're likely to need support, best to wait for the proper release, it shouldn't be long.
